### ISSUE TITLE
requestLocation fails

### ISSUE THREAD

**Issue by joshbartley** [stripe/stripe-terminal-ios#56]
Working on getting the iOS side complete for our Xamarin app. I can discover readers, connect, check status, see my terminal token client being called, all of that is working. When I try to call ProcessPayment I never get a call back to my method.

I enabled verbose logging, after calling ProcessPayment these lines are emitted. Removed the IDs from the events even though this was all the test keys and simulated reader. I can see the Payment Intent is set to a Requires Confirmation, but I don't see the /confirm call in the dashboard like I would for the Android side to know it was able to process the payment. 

> command=processPayment event=processPayment_completion_start intent_status=requires_confirmation last_request_id=req_XXXXXXXXXXXX 
> command_in_progress=processPayment event=state_change last_request_id=req_XXXXXXXXXX
> event=state_change last_request_id=req_XXXXXXXX logpoint_level=info payment_intent=pi_XXXXXXXXX payment_status=Processing 
> event=processPayment_completion last_request_id=req_XXXXXXX logpoint_level=info payment_intent=pi_XXXXXXXXX payment_intent_status=requires_confirmation reader=rdr_XXXXXXXXXXXXXXXXX scope=SCPPaymentSession sdk_version=1.0.0
> event=requestLocation last_request_id=req_XXXXXXXXXXXX
> 

The Payment Intent ID and Request ID match up to what I see in the Stripe dashboard. Location services is enabled for the simulator, Location access is enabled for the app, and Location When In Use is in the Info.plist. I looked into location since that was the last call. Nothing else is emitted from the logs after the `event=requestLocation` event.

**Comment by joshbartley** [comment]
There is some sort of threading issue I'm running into. If I call `processPayment` for a second time, I get the error that the SDK is busy with 'processPayment'. I had this happen with the fetch token, and made the it synchronous, which resolved the issue. My processPayment flow is already synchronous. I can't look at the code to debug this further since I can't see what processPayment does internally or how it handles threads.

**Comment by joshbartley** [comment]
I've narrowed this down to the `StripeTerminal.framework` library can't access location data. My app can request the location data directly and that works just fine. I realized this when the DiscoverReaders call also asks for the location and gets a return event that the location was updated. While in my app this never happens. The info.plist is set correctly and I am tried targeting both the 9.0 and 12.0 SDKs and both are the same result.

If I go into the Privacy menu and deny access to location to the app, I get a response on the ProcessPayment callback that location services are disabled and are required.

It would be helpful to know more about how the library is trying to get the location so I can setup a test to make sure that iOS call is not getting left out from a linker or some other process.

**Comment by joshbartley** [comment]
After debugging this, the issue was an undocumented requirement for the SDK. CLLocationManager can only be requested in an active RunLoop thread. In the example, everything is in the viewDidLoad function, which meets that requirement. If you don't call into the SDK with an active RunLoop thread, the SDK hangs waiting on a location because itself didn't protect against the bug, just assumed it would be called. I figured this out once i looked the Xamarin.Essentials GeoLocation function for iOS that forces a CLLocationManager off the correct thread.

I'm leaving this open as an issue to address the documentation around this as the Stripe Terminal Getting Started documentation is not included in this repo.

**Comment by calhoun137** [comment]
Thanks for making this issue josh, this was very helpful to our team.  We faced the exact same issue, we are using React Native, not sure if you are or not.  I have found the solution for our React Native app, and am leaving this here in case it helps anyone else.  I have been in touch with stripe support and they let us know they are looking in to this and may put out an update with a fix.

```
#import <React/RCTEventDispatcher.h> // NOTE: this may not be required, not sure
...
- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}
```

**Comment by calhoun137** [comment]
to further clarify, what I learned is that there are multiple "queues" in iOS and react native keeps module code in some other queue besides the one required for the main thread only ios-API.  We were in the main thread of the wrong queue, but using the interface method above, were able to get the callback to return normally like it should.

**Comment by joshbartley** [comment]
@calhoun137 when I was debugging this I saw that the connectReader call tries to get the location as well. So you may want to put that call in the main queue as well in case there are any side effects to a hanging CLLocationManager request on the Stripe side. To be save I just wrapped call calls into the stripe terminal to the main thread, including the initTerminal as the callback for the token provider was acting weird a few times. I was using Xamarin/C# for my code which the threads behave similar to the react native since it doesn't all run under the main thread in iOS.

**Comment by bg-stripe** [comment]
@calhoun137 Thanks for sharing your progress and collaborating on a solution! 

In case you didn't know about this one, a few users have been collaborating on this React Native project: https://github.com/theopolisme/react-native-stripe-terminal

Until we have official support, it would be really great to see users collaborate on a few consolidated open source projects. We can also do a better job linking to these options from our Readmes and docs.

**Comment by calhoun137** [comment]
@bg-stripe Thanks!  It would be super fun to contribute to that project and we are strongly considering it.  After taking a quick look I noticed there is no Android support, we are already quite far along and perhaps that is something we can contribute in the near future.

I think the expected behavior is that if a function call like readReusableCard fails because it was called from the wrong thread, that the callback should return with an error instead of hanging.  This would be my suggestion of how to address this issue in an update.

**Comment by bric-stripe** [comment]
👋 I missed that this issue existed and just realized this matches an issue we fixed in [2.2.0](https://github.com/stripe/stripe-terminal-ios/releases/tag/v2.2.0):

> Bug fix: Fixed a possible deadlock when calling readReusableCard and processPayment from background threads.

I realize folks have likely sorted things out but closing out this old issue as that should no longer get the SDK stuck
